src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,182 @@
+                }
+                tempNode = tempNode.prev;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForBound(index);
+        return getNode(index).body;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForBound(index);
+        T previousBody = getNode(index).body;
+        getNode(index).body = value;
+        return previousBody;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> toRemoveNode = getNode(index);
+
+        unlinkNode(toRemoveNode);
+        size--;
+        return toRemoveNode.body;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (get(i) == object || get(i) != null && get(i).equals(object)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkNewNode(Node<T> tempNode, T value) {
+        Node<T> toAdd = new Node<>(tempNode.prev, value, tempNode);
+        tempNode.prev.next = toAdd;
+        tempNode.prev = toAdd;
+        size++;
+    }
+
+    private void checkIndexForBound(int index) {
+        String exception = String.format("Wrong index! %d index is out of bounds.", index);
+        if (index < 0 || index > size - 1 && size > 1) {
+            throw new IndexOutOfBoundsException(exception);
+        }
+    }
+
+    private void checkIndexForBoundAdd(int index) {
+        String exception = String.format("Wrong index! %d index is out of bounds.", index);
+        if (index < 0 || index > size && size > 1) {
+            throw new IndexOutOfBoundsException(exception);
+        }
+    }
+
+    private void unlinkNode(Node<T> node) {
+        Node<T> nextNode = node.next;
+        Node<T> previousNode = node.prev;
+        if (node == head) {
+            head = nextNode;
+        } else if (node == tail) {
+            tail = previousNode;
+        }
+        if (previousNode != null) {
+            previousNode.next = nextNode;
+        }
+        if (nextNode != null) {
+            nextNode.prev = previousNode;
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndexForBound(index);
+        int firstHalf = size / 2;
+        Node<T> correctNode = null;
+        if (index <= firstHalf) {
+            Node<T> tempNode = head;
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    correctNode = tempNode;
+                }
+                tempNode = tempNode.next;
+            }
```

Comment:
you don't need to iterate over the list. just call 'index' times tempNode = tempNode.next and you'll have your node